### ISSUE TITLE
kube-apiserver unable to write to etcd

### ISSUE THREAD

**Issue by cjcullen** [GoogleCloudPlatform/kubernetes#10122]
In about 100 test runs overnight, 5 clusters failed with the same problem:
Nodes try to register with the apiserver, apiserver sends back a 500.
Looking at the kube-apiserver logs, I see:
```
E0619 20:37:50.685032       5 errors.go:62] apiserver received an error that is not an api.Status: 501: All the given peers are not reachable (failed to propose on members [http://127.0.0.1:4001] twice [last error: Unexp
ected HTTP status code]) [0]
I0619 20:37:50.685229       5 handlers.go:135] POST /api/v1/namespaces: (12.228060557s) 500
...
E0619 20:37:50.643558       5 errors.go:62] apiserver received an error that is not an api.Status: 501: All the given peers are not reachable (failed to propose on members [http://127.0.0.1:4001] twice [last error: Unexp
ected HTTP status code]) [0]
I0619 20:37:50.643762       5 handlers.go:135] POST /api/v1/nodes: (12.228247163s) 500
```

To me, that looks like writes to etcd are failing. Looking into the etcd container logs, I see a lot of:

```
2015/06/19 20:40:38 etcdhttp: unexpected error: etcdserver: request timed out
2015/06/19 20:40:38 etcdhttp: unexpected error: etcdserver: request timed out
2015/06/19 20:40:38 etcdserver: publish error: etcdserver: request timed out
2015/06/19 20:40:38 etcdhttp: unexpected error: etcdserver: request timed out
2015/06/19 20:40:41 etcdhttp: unexpected error: etcdserver: request timed out
2015/06/19 20:40:41 etcdhttp: unexpected error: etcdserver: request timed out
2015/06/19 20:40:42 etcdhttp: unexpected error: etcdserver: request timed out
2015/06/19 20:40:43 etcdhttp: unexpected error: etcdserver: request timed out
2015/06/19 20:40:43 etcdhttp: unexpected error: etcdserver: request timed out
2015/06/19 20:40:43 etcdhttp: unexpected error: etcdserver: request timed out
2015/06/19 20:40:43 etcdserver: publish error: etcdserver: request timed out
2015/06/19 20:40:44 etcdhttp: unexpected error: etcdserver: request timed out
2015/06/19 20:40:44 etcdhttp: unexpected error: etcdserver: request timed out
2015/06/19 20:40:44 etcdhttp: unexpected error: etcdserver: request timed out
2015/06/19 20:40:47 etcdhttp: unexpected error: etcdserver: request timed out
2015/06/19 20:40:47 etcdhttp: unexpected error: etcdserver: request timed out
2015/06/19 20:40:47 etcdhttp: unexpected error: etcdserver: request timed out
```

Trying to hit etcd directly:
```
# curl -L http://127.0.0.1:4001/v2/keys/message -XPUT -d value="Hello world"
{"message":"Internal Server Error"}
```
Etcd's data directory is pretty empty, other than one entry in members/wal, so I tried listing members:
```
# curl http://127.0.0.1:4001/v2/members
{"members":[]}
```
In a working cluster, that looks more like:
```
# curl http://127.0.0.1:4001/v2/members
{"members":[{"id":"ce2a822cea30bfca","name":"default","peerURLs":["http://localhost:2380","http://localhost:7001"],"clientURLs":["http://127.0.0.1:4001"]}]}
```

I'm about at the extent of my etcd knowledge, so if anybody has any idea what else to poke at, I'd love to hear it.

**Comment by cjcullen** [comment]
@davidopp @lavalamp @alex-mohr

**Comment by cjcullen** [comment]
Comparing the startup of a functioning etcd container to my non-functioning one:

**Functional**
```
2015/06/19 21:24:24 etcd: already initialized as member before, starting as etcd member...
2015/06/19 21:24:24 etcd: listening for peers on http://localhost:2380
2015/06/19 21:24:24 etcd: listening for peers on http://localhost:7001
2015/06/19 21:24:24 etcd: listening for client requests on http://127.0.0.1:4001
2015/06/19 21:24:24 etcdserver: datadir is valid for the 2.0.1 format
2015/06/19 21:24:24 etcdserver: recovered store from snapshot at index 440044
2015/06/19 21:24:24 etcdserver: name = default
2015/06/19 21:24:24 etcdserver: data dir = /var/etcd/data
2015/06/19 21:24:24 etcdserver: member dir = /var/etcd/data/member
2015/06/19 21:24:24 etcdserver: heartbeat = 100ms
2015/06/19 21:24:24 etcdserver: election = 1000ms
2015/06/19 21:24:24 etcdserver: snapshot count = 10000
2015/06/19 21:24:24 etcdserver: advertise client URLs = http://127.0.0.1:4001
2015/06/19 21:24:24 etcdserver: loaded cluster information from store: default=http://localhost:2380,default=http://localhost:7001
2015/06/19 21:24:24 etcdserver: restart member ce2a822cea30bfca in cluster 7e27652122e8b2ae at commit index 448981
2015/06/19 21:24:24 raft: ce2a822cea30bfca became follower at term 2
2015/06/19 21:24:24 raft: newRaft ce2a822cea30bfca [peers: [ce2a822cea30bfca], term: 2, commit: 448981, applied: 440044, lastindex: 448981, lastterm: 2]
2015/06/19 21:24:26 raft: ce2a822cea30bfca is starting a new election at term 2
2015/06/19 21:24:26 raft: ce2a822cea30bfca became candidate at term 3
2015/06/19 21:24:26 raft: ce2a822cea30bfca received vote from ce2a822cea30bfca at term 3
2015/06/19 21:24:26 raft: ce2a822cea30bfca became leader at term 3
2015/06/19 21:24:26 raft.node: ce2a822cea30bfca elected leader ce2a822cea30bfca at term 3
2015/06/19 21:24:26 etcdserver: published {Name:default ClientURLs:[http://127.0.0.1:4001]} to cluster 7e27652122e8b2ae
```
**Non-functional**
```
2015/06/19 21:21:29 etcd: already initialized as member before, starting as etcd member...
2015/06/19 21:21:29 etcd: listening for peers on http://localhost:2380
2015/06/19 21:21:29 etcd: listening for peers on http://localhost:7001
2015/06/19 21:21:29 etcd: listening for client requests on http://127.0.0.1:4001
2015/06/19 21:21:29 etcdserver: datadir is valid for the 2.0.1 format
2015/06/19 21:21:29 etcdserver: name = default
2015/06/19 21:21:29 etcdserver: data dir = /var/etcd/data
2015/06/19 21:21:29 etcdserver: member dir = /var/etcd/data/member
2015/06/19 21:21:29 etcdserver: heartbeat = 100ms
2015/06/19 21:21:29 etcdserver: election = 1000ms
2015/06/19 21:21:29 etcdserver: snapshot count = 10000
2015/06/19 21:21:29 etcdserver: advertise client URLs = http://127.0.0.1:4001
2015/06/19 21:21:29 etcdserver: restart member ce2a822cea30bfca in cluster 7e27652122e8b2ae at commit index 0
2015/06/19 21:21:29 raft: ce2a822cea30bfca became follower at term 0
2015/06/19 21:21:29 raft: newRaft ce2a822cea30bfca [peers: [], term: 0, commit: 0, applied: 0, lastindex: 1, lastterm: 1]
```

It looks like the lone etcd server is choosing to become a follower of nothing?

**Comment by cjcullen** [comment]
On my non-functional master, it looks like the etcd container came up and was on its way to working when it was torn down:
```
# docker logs 87e75f1b9276
2015/06/19 06:33:42 etcd: listening for peers on http://localhost:2380
2015/06/19 06:33:42 etcd: listening for peers on http://localhost:7001
2015/06/19 06:33:42 etcd: listening for client requests on http://127.0.0.1:4001
2015/06/19 06:33:42 etcdserver: datadir is valid for the 2.0.1 format
2015/06/19 06:33:42 etcdserver: name = default
2015/06/19 06:33:42 etcdserver: data dir = /var/etcd/data
2015/06/19 06:33:42 etcdserver: member dir = /var/etcd/data/member
2015/06/19 06:33:42 etcdserver: heartbeat = 100ms
2015/06/19 06:33:42 etcdserver: election = 1000ms
2015/06/19 06:33:42 etcdserver: snapshot count = 10000
2015/06/19 06:33:42 etcdserver: advertise client URLs = http://127.0.0.1:4001
2015/06/19 06:33:42 etcdserver: initial advertise peer URLs = http://localhost:2380,http://localhost:7001
2015/06/19 06:33:42 etcdserver: initial cluster = default=http://localhost:2380,default=http://localhost:7001
2015/06/19 06:33:42 etcdserver: start member ce2a822cea30bfca in cluster 7e27652122e8b2ae
2015/06/19 06:33:42 raft: ce2a822cea30bfca became follower at term 0
2015/06/19 06:33:42 raft: newRaft ce2a822cea30bfca [peers: [], term: 0, commit: 0, applied: 0, lastindex: 0, lastterm: 0]
2015/06/19 06:33:42 raft: ce2a822cea30bfca became follower at term 1
2015/06/19 06:33:42 etcdserver: added local member ce2a822cea30bfca [http://localhost:2380 http://localhost:7001] to cluster 7e27652122e8b2ae
2015/06/19 06:33:43 received terminated signal, shutting down...
2015/06/19 06:33:42 etcd: listening for peers on http://localhost:2380
2015/06/19 06:33:42 etcd: listening for peers on http://localhost:7001
2015/06/19 06:33:42 etcd: listening for client requests on http://127.0.0.1:4001
2015/06/19 06:33:42 etcdserver: datadir is valid for the 2.0.1 format
2015/06/19 06:33:42 etcdserver: name = default
2015/06/19 06:33:42 etcdserver: data dir = /var/etcd/data
2015/06/19 06:33:42 etcdserver: member dir = /var/etcd/data/member
2015/06/19 06:33:42 etcdserver: heartbeat = 100ms
2015/06/19 06:33:42 etcdserver: election = 1000ms
2015/06/19 06:33:42 etcdserver: snapshot count = 10000
2015/06/19 06:33:42 etcdserver: advertise client URLs = http://127.0.0.1:4001
2015/06/19 06:33:42 etcdserver: initial advertise peer URLs = http://localhost:2380,http://localhost:7001
2015/06/19 06:33:42 etcdserver: initial cluster = default=http://localhost:2380,default=http://localhost:7001
2015/06/19 06:33:42 etcdserver: start member ce2a822cea30bfca in cluster 7e27652122e8b2ae
2015/06/19 06:33:42 raft: ce2a822cea30bfca became follower at term 0
2015/06/19 06:33:42 raft: newRaft ce2a822cea30bfca [peers: [], term: 0, commit: 0, applied: 0, lastindex: 0, lastterm: 0]
2015/06/19 06:33:42 raft: ce2a822cea30bfca became follower at term 1
2015/06/19 06:33:42 etcdserver: added local member ce2a822cea30bfca [http://localhost:2380 http://localhost:7001] to cluster 7e27652122e8b2ae
2015/06/19 06:33:43 received terminated signal, shutting down...
```

**Comment by cjcullen** [comment]
Looks like this is a known [etcd bug](https://github.com/coreos/etcd/pull/2878), fixed in 2.0.12.

The workaround is "Do not shutdown the etcd member too fast (wait for at least a second)." Now to figure out why we are killing the etcd container...

**Comment by dchen1107** [comment]
I read [etcd bug](https://github.com/coreos/etcd/pull/2878), and this issue. Here is my question: 

1) Who kill etcd at the first place? kubelet? because liveness failure? 
2) If liveness is the case, we might misconfigure the Probe with not long enough InitialDelaySeconds?

**Comment by lavalamp** [comment]
No, it's the old kubelet restarts docker thing. :/

**Comment by dchen1107** [comment]
Ahh, another chicken-egg issue at bootstrap time.

**Comment by cjcullen** [comment]
In this case, I don't think it's kubelet killing the pod. I think it is salt bouncing docker after its initial configuration of cbr0.

**Comment by cjcullen** [comment]
Looks like updating to etcd 2.0.12 did not fix this issue. Another failure from last night:
From the kube-apiserver logs.
```
I0622 20:34:03.081625       5 trace.go:57] Trace "Create *api.Namespace" (started 2015-06-22 20:33:50.854808199 +0000 UTC):
[53.755µs] [53.755µs] About to create object
[12.226780329s] [12.226726574s] END
I0622 20:34:03.081625       5 trace.go:57] Trace "Create *api.Namespace" (started 2015-06-22 20:33:50.853932917 +0000 UTC):
[84.009µs] [84.009µs] About to create object
[12.227655603s] [12.227571594s] END
E0622 20:34:03.081680       5 errors.go:62] apiserver received an error that is not an api.Status: 501: All the given peers are not reachable (failed to propose on members [http://127.0.0.1:4001] twice [last error: Unexpected HTTP status code]) [0]
E0622 20:34:03.081686       5 errors.go:62] apiserver received an error that is not an api.Status: 501: All the given peers are not reachable (failed to propose on members [http://127.0.0.1:4001] twice [last error: Unexpected HTTP status code]) [0]
```
And in the etcd container's logs:
```
...
2015/06/22 16:18:48 etcdhttp: unexpected error: etcdserver: request timed out
2015/06/22 16:18:48 etcdserver: publish error: etcdserver: request timed out
2015/06/22 16:18:51 etcdhttp: unexpected error: etcdserver: request timed out
2015/06/22 16:18:51 etcdhttp: unexpected error: etcdserver: request timed out
2015/06/22 16:18:51 etcdhttp: unexpected error: etcdserver: request timed out
2015/06/22 16:18:52 etcdhttp: unexpected error: etcdserver: request timed out
...
```
Just to confirm:
```
# curl "http://127.0.0.1:4001/version"
etcd 2.0.12
```

**Comment by dchen1107** [comment]
Ok, I have a fix based on #10153 which doesn't execute any podspecs until network configuration is done no matter it is master nodes or minion nodes, the detail can be found at  https://github.com/GoogleCloudPlatform/kubernetes/pull/10153#issuecomment-113729964. Then I start validating the fix. I never run into the issue reported here and #9822 even without the fix. So to make sure no race completely between configing cbr0 and starting static pods, there shouldn't have any master component related containers being killed by docker at booting time. But unfortunately, I still observed a lot killings which could cause etcd_server failure reported here, or #9822 or some other unknown issues for master components. Below is one of those examples:

```
# docker ps -a
...
4bfdec46289e        gcr.io/google_containers/kube-apiserver:d66676aac636c00aff5833d555a27e6f            "/bin/sh -c '/usr/lo   30 minutes ago      Exited (0) 30 minutes ago                       k8s_kube-apiserver.2060aa6b_kube-apiserver-kubernetes-master_default_9b28a43ce4b81f8ef3aa697738b5b009_e23fa300                    
...
```

To make sure kubelet does configure cbr0 and restart docker
```
# cat kubelet.log | grep "Recreated cbr0 and restarted docker"
I0623 00:23:26.802295    3536 container_bridge.go:51] Recreated cbr0 and restarted docker
```

CBR0 does configured properly, and docker is restarted due to it. Also syncLoop does skipped if network is not configured properly. Then why? 

Then I checked the timestamps of docker container, and found it is killed before the new version of kubelet run, which means it is started by pre-installed kubelet and killed by docker later: 

```
# docker inspect 4bfdec46289e

    "State": {
        "Dead": false,
        "Error": "",
        "ExitCode": 0,
        "FinishedAt": "2015-06-23T00:23:26.889106704Z",       <======================= Note the timestamp here is around the same time of above kubelet log. 
        "OOMKilled": false,
        "Paused": false,
        "Pid": 0,
        "Restarting": false,
        "Running": false,
        "StartedAt": "2015-06-23T00:23:21.647061431Z"
    },
```

Kubelet log also proves this:

```
# cat kubelet.log | grep  4bfdec46289e
I0623 00:23:21.361813    2368 event.go:203] Event(api.ObjectReference{Kind:"Pod", Namespace:"default", Name:"kube-apiserver-kubernetes-master", UID:"9b28a43ce4b81f8ef3aa697738b5b009", APIVersion:"v1", ResourceVersion:"", FieldPath:"spec.containers{kube-apiserver}"}): reason: 'created' Created with docker id 4bfdec46289e34ede91ff8a5d87e7857586b3b6078b4e0000decaab0771aa54a
I0623 00:23:21.678145    2368 event.go:203] Event(api.ObjectReference{Kind:"Pod", Namespace:"default", Name:"kube-apiserver-kubernetes-master", UID:"9b28a43ce4b81f8ef3aa697738b5b009", APIVersion:"v1", ResourceVersion:"", FieldPath:"spec.containers{kube-apiserver}"}): reason: 'started' Started with docker id 4bfdec46289e34ede91ff8a5d87e7857586b3b6078b4e0000decaab0771aa54a
```

After digging into it more, I realized that we are going to have this race until re-cutting containervm image with the new fix I had or I have to make sure kube-master-addons run after newly installed kubelet and docker (only flags) run. Based on my testing of shuting down cluster and recreating, actually this race causes most the killings at the boot time. Thus, I introduced https://github.com/dchen1107/kubernetes-1/commit/eb3c8ae92c5cbd7a5ab8a097a805302465ca16c2 to handle this.
